src/main/java/org/folio/circulation/services/ActualCostRecordService.java

```diff
@@ -0,0 +1,82 @@
+import static org.folio.circulation.services.LostItemFeeChargingService.ReferenceDataContext;
+import static org.folio.circulation.support.results.Result.succeeded;
+import static org.folio.circulation.support.results.ResultBinding.mapResult;
+public class ActualCostRecordService {
+  private final ActualCostRecordRepository actualCostStorageRepository;
+
+  public ActualCostRecordService(ActualCostRecordRepository actualCostStorageRepository) {
+    this.actualCostStorageRepository = actualCostStorageRepository;
+  }
+
+  public CompletableFuture<Result<ReferenceDataContext>> createActualCostRecordIfNecessary(
+    ReferenceDataContext referenceDataContext) {
+    return createActualCostRecordIfNecessary(referenceDataContext.getLoan(),
+      referenceDataContext.getFeeFineOwner(), ItemLossType.DECLARED_LOST, ZonedDateTime.now(), referenceDataContext.getFeeFines().stream()
+        .filter(feeFine -> LOST_ITEM_ACTUAL_COST_FEE_TYPE.equals(feeFine.getFeeFineType()))
+        .findFirst().orElse(null))
+      .thenApply(mapResult(referenceDataContext::withActualCostRecord));
+  }
+
+  public CompletableFuture<Result<LoanToChargeFees>> createActualCostRecordIfNecessary(
+    LoanToChargeFees loanToChargeFees) {
+
+    Loan loan = loanToChargeFees.getLoan();
+    return createActualCostRecordIfNecessary(loan, loanToChargeFees.getOwner(), ItemLossType.AGED_TO_LOST,
+      loan.getAgedToLostDateTime(), loanToChargeFees.getFeeFineTypes().get(LOST_ITEM_ACTUAL_COST_FEE_TYPE))
+      .thenApply(mapResult(loanToChargeFees::withActualCostRecord));
+  }
+
+  public CompletableFuture<Result<ActualCostRecord>> createActualCostRecordIfNecessary(Loan loan,
+    FeeFineOwner feeFineOwner, ItemLossType lossType, ZonedDateTime dateOfLoss, FeeFine actualCostFeeFine) {
+
+    return loan.getLostItemPolicy().hasActualCostFee() ?
+      actualCostStorageRepository.createActualCostRecord(buildActualCostRecord(loan, feeFineOwner,
+        lossType, dateOfLoss, actualCostFeeFine)) :
+      completedFuture(succeeded(null));
+  }
+
+  private ActualCostRecord buildActualCostRecord(Loan loan, FeeFineOwner feeFineOwner,
+    ItemLossType lossType, ZonedDateTime dateOfLoss, FeeFine actualCostFeeFine) {
+
+    Item item = loan.getItem();
+
+    Location permanentLocation = item.getPermanentLocation();
+    return new ActualCostRecord()
+      .withUserId(loan.getUserId())
+      .withUserBarcode(loan.getUser().getBarcode())
+      .withLoanId(loan.getId())
+      .withItemLossType(lossType)
+      .withDateOfLoss(dateOfLoss.toString())
+      .withTitle(item.getTitle())
+      .withIdentifiers(item.getIdentifiers()
+        .collect(Collectors.toList()))
+      .withItemBarcode(item.getBarcode())
+      .withLoanType(item.getLoanTypeName())
+      .withCallNumberComponents(item.getCallNumberComponents())
+      .withPermanentItemLocation(permanentLocation == null || permanentLocation.getName() == null ?
+        "" : permanentLocation.getName())
```

Comment:
This condition I also was requested to move to a method and put it to Item. Pls, check..

Reply:
Extracted